WtgPlugin/src/org/workcraft/plugins/wtg/utils/VerificationUtils.java

```diff
@@ -872,4 +873,274 @@ private static boolean isFirstTransition(Wtg wtg, MathNode node) {
+            firedAfterOut.put(waveform, firedAfter);
+            hasOutput.put(waveform, hasOutputTransition(wtg, waveform));
+        }
+
+        Set<String> inputSignals = new HashSet<>();
+        for (Signal signal : wtg.getSignals()) {
+            if (signal.getType() == Signal.Type.INPUT) {
+                inputSignals.add(wtg.getName(signal));
+            }
+        }
+
+        for (String signal : inputSignals) {
+            if (!signalFollowsSynthesisGuidelines(wtg, signal, firedBeforeOut, firedAfterOut, hasOutput)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private static boolean signalFollowsSynthesisGuidelines(Wtg wtg, String signal, Map<Waveform, Set<String>> firedBeforeOut,
+                                                            Map<Waveform, Set<String>> firedAfterOut, Map<Waveform, Boolean> hasOutput) {
+        Queue<Waveform> toVisit = new LinkedList<>();
+        Set<Waveform> visited = new HashSet<>();
+        //We only need to look waveforms in which the signal fired after any output
+        for (Map.Entry<Waveform, Set<String>> entry : firedAfterOut.entrySet()) {
+            if (entry.getValue().contains(signal)) {
+                toVisit.add(entry.getKey());
+                visited.add(entry.getKey());
+            }
+        }
+
+        while (!toVisit.isEmpty()) {
+            Waveform visitingWaveform = toVisit.poll();
+
+            State followingState;
+            if ((wtg.getPostset(visitingWaveform).iterator().hasNext()) &&
+                    (wtg.getPostset(visitingWaveform).iterator().next() instanceof State)) {
+                followingState = (State) wtg.getPostset(visitingWaveform).iterator().next();
+            } else {
+                continue;
+            }
+
+            for (MathNode node : wtg.getPostset(followingState)) {
+                if (!(node instanceof Waveform)) {
+                    continue;
+                }
+                Waveform nextWaveform = (Waveform) node;
+
+                if (firedBeforeOut.get(nextWaveform).contains(signal)) {
+                    DialogUtils.showError("Input signal '" + signal + "' can fire before reaching waveform '" +
+                            wtg.getName(nextWaveform) +
+                            "' without an output in between, yet it can fire in that waveform before any output does.");
+                    return false;
+                }
+
+                //If a succeeding waveform does not have any output, we have to look at its succeeding waveforms too
+                if ((!hasOutput.get(nextWaveform)) && (!visited.contains(nextWaveform))) {
+                    toVisit.add(nextWaveform);
+                    visited.add(nextWaveform);
+                }
+            }
+        }
+        return true;
+    }
+
+
+    private static boolean waveformFollowsSynthesisGuidelines(Wtg wtg, Waveform waveform, Set<String> firedBeforeOut,
```

Comment:
Is it possible to refactor pieces of this function out, so it is a bit more compact?

Reply:
I updated my repository with a refactored version of that function. It's still a bit big. If you think it's too big I can package more parts of the core into smaller functions.

Reply:
It is definitely better, let's keep it as is for now.